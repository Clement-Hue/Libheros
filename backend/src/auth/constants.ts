export const jwtSecret =
  '38647c567b3ad28e760746352d676bd2aa493aaedd375d6d572e310210c0f4a1173119d62a3b49b534fdcce670cfb69ecad498caf478a1a0a66972b8b9e7006bdc9938417319a7b23308218edd378fae21bce468e29195c9d4e5ee7dd9adfcbdfc05b39c1282a81b6375b5643c33467adb141f1a7ffdf42c95574a1ee63a54a526ab9795b8560542184de367c4baa89e5d83e36b64b7afe84cb2d38bff71a808c8e60ec89a7d28b8b02d294103692f18b55d5a8d26f71e8d9af5a5d08aafc6f825e451eb5fe2a256a00b1ba8c6e4039667ee861e12ba3ab1565ffed66a8cc7f4c5118509f1775454bf023a3a6a36652fcd6275ef277540fb4cac5beddd307146';

