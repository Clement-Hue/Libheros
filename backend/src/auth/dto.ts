import { IsEmail, IsNotEmpty } from 'class-validator';

export class AuthDto {
  @IsNotEmpty()
  @IsEmail()
  email: string;

  @IsNotEmpty()
  password: string;
}

export class CreateUserDto {
  @IsNotEmpty()
  @IsEmail()
  email: string;
  @IsNotEmpty()
  firstName: string;
  @IsNotEmpty()
  lastName: string;
  @IsNotEmpty()
  password: string;
}
