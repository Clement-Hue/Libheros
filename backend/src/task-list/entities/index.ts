export { default as Task } from './task.entity';
export { default as TaskList } from './taskList.entity';
