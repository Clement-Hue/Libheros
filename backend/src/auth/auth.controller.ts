import { Body, Controller, HttpCode, HttpStatus, Post } from '@nestjs/common';
import { AuthService } from './auth.service';
import { CreateUserDto } from './dto';

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('register')
  @HttpCode(HttpStatus.CREATED)
  async register(@Body() data: CreateUserDto) {
    await this.authService.createUser({...data, taskLists: []});
  }
}
