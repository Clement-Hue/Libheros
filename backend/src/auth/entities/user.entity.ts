import { TaskList } from '../../task-list/entities';
import { Entity, Column, PrimaryColumn, OneToMany } from 'typeorm';

@Entity()
export default class User {
  @PrimaryColumn()
  email: string;

  @Column()
  firstName: string;

  @Column()
  lastName: string;

  @Column()
  password: string;

  @OneToMany(() => TaskList, (taskList) => taskList.user)
  taskLists: TaskList[];
}
