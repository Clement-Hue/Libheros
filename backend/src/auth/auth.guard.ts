import {
  CanActivate,
  ExecutionContext,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { jwtSecret } from './constants';
import { Request } from 'express';
import { Repository } from 'typeorm';
import { User } from './entities';
import { InjectRepository } from '@nestjs/typeorm';

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(
    private jwtService: JwtService,
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) {}

  async canActivate(context: ExecutionContext): Promise<boolean> {
    const request = context.switchToHttp().getRequest();
    const token = this.extractTokenFromHeader(request);
    if (!token) {
      throw new UnauthorizedException();
    }
    try {
      const payload = await this.jwtService.verifyAsync(token, {
        secret: jwtSecret,
      });

      request['user'] = await this.userRepository.findOneOrFail({
        where: { email: payload.sub },
      });
    } catch {
      throw new UnauthorizedException({ code: 'auth.invalid-token' });
    }
    return true;
  }

  private extractTokenFromHeader(request: Request): string | undefined {
    const [type, token] = request.headers.authorization?.split(' ') ?? [];
    return type === 'Bearer' ? token : undefined;
  }
}
