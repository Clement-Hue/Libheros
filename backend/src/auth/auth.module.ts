import { Module } from '@nestjs/common';
import { AuthController } from './auth.controller';
import { TypeOrmModule } from '@nestjs/typeorm';
import { User } from './entities';
import { AuthService } from './auth.service';
import { JwtModule } from '@nestjs/jwt';
import { jwtSecret } from './constants';

const TypeOrmUser = TypeOrmModule.forFeature([User])

@Module({
  imports: [
    TypeOrmUser,
    JwtModule.register({
      global: true,
      secret: jwtSecret,
      signOptions: { expiresIn: '2h' },
    }),
  ],
  controllers: [AuthController],
  providers: [AuthService],
  exports: [TypeOrmUser],
})
export class AuthModule {}
