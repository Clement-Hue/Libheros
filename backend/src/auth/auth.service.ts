import { Injectable, UnauthorizedException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import * as bcrypt from 'bcrypt';
import { Repository } from 'typeorm';
import { User } from './entities';

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) {}

  async createUser(data: User): Promise<User> {
    const saltRounds = 10;
    const hashedPassword = await bcrypt.hash(data.password, saltRounds);
    const user = this.userRepository.create({
      ...data,
      password: hashedPassword,
    });
    return this.userRepository.save(user);
  }

  async auth(email: string, password: string) {
    const user = await this.userRepository.findOne({ where: { email } });
    if (!user) {
      throw new UnauthorizedException({code: "auth.invalid-credentials"});
    }
    const isMatch = await bcrypt.compare(password, user.password);
    if (!isMatch) {
      throw new UnauthorizedException({ code: 'auth.invalid-credentials' });
    }
  }
}
