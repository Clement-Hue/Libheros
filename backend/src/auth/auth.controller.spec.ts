import { Test, TestingModule } from '@nestjs/testing';
import { AuthController } from './auth.controller';
import { INestApplication } from '@nestjs/common';
import { Repository } from 'typeorm';
import {
  createTestingApp,
  TestingModules,
} from '../test-utils/testing-modules';
import { getRepositoryToken } from '@nestjs/typeorm';
import { User } from './entities';
import { AuthService } from './auth.service';
import * as request from 'supertest';
import { makeUser } from '../test-utils/factories';
import { JwtService } from '@nestjs/jwt';
import { jwtSecret } from './constants';
import * as bcrypt from "bcrypt"

describe('AuthController', () => {
  let app: INestApplication;
  let userRepository: Repository<User>;
  let jwtService: JwtService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: TestingModules(),
      controllers: [AuthController],
      providers: [AuthService],
    }).compile();

    app = await createTestingApp(module);
    userRepository = module.get<Repository<User>>(getRepositoryToken(User));
    jwtService = module.get<JwtService>(JwtService);
  });

  afterAll(async () => {
    await app.close();
  });

  it('should create a user', async () => {
    const payload = {
      firstName: 'Pierre',
      lastName: 'Dupont',
      email: 'pierre@gmail.com',
      password: 'password123',
    };
    await request(app.getHttpServer())
      .post('/auth/register')
      .send(payload)
      .expect(201);
    expect(
      await userRepository.findOne({ where: { email: payload.email } }),
    ).toMatchSnapshot({
      password: expect.any(String),
    });
  });

  it("should get access token", async () => {
    const payload = {
      email: 'pierre@gmail.com',
      password: 'password123',
    };
    const user = makeUser({
      email: payload.email,
      password: bcrypt.hashSync(payload.password, 10)
    })
    await userRepository.save(user)
    const res = await request(app.getHttpServer())
      .post('/auth')
      .send(payload)
      .expect(200);
    const jwtPayload = jwtService.verify(res.body.data?.access_token, {secret: jwtSecret})
    expect(jwtPayload.sub).toEqual(payload.email)
  })

  it("should show error invalid password", async () => {
    const payload = {
      email: 'pierre@gmail.com',
      password: 'pass',
    };
    const user = makeUser({
      email: payload.email,
      password: bcrypt.hashSync("password123!", 10)
    })
    await userRepository.save(user)
    const res = await request(app.getHttpServer())
      .post('/auth')
      .send(payload)
      .expect(401);
    expect(res.body).toMatchSnapshot()
  })

  it("should show error invalid email", async () => {
    const payload = {
      email: 'wrong@gmail.com',
      password: 'password123!',
    };
    const user = makeUser({
      email: 'pierre@gmail.com',
      password: bcrypt.hashSync(payload.password, 10)
    })
    await userRepository.save(user)
    const res = await request(app.getHttpServer())
      .post('/auth')
      .send(payload)
      .expect(401);
    expect(res.body).toMatchSnapshot()
  })
});
