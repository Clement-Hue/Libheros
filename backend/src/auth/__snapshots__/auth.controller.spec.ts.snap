// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthController should create a user 1`] = `
{
  "email": "pierre@gmail.com",
  "firstName": "Pierre",
  "lastName": "Dupont",
  "password": Any<String>,
}
`;
